handout/testing/services/mockbackend.md

```diff
@@ -0,0 +1,103 @@
+
+*wikisearch.spec.ts*
+
+```js
+import {
+  fakeAsync,
+  inject,
+  TestBed
+} from '@angular/core/testing';
+import {
+  HttpModule,
+  XHRBackend,
+  ResponseOptions,
+  Response
+} from '@angular/http';
+import {
+  MockBackend,
+  MockConnection
+} from '@angular/http/testing/mock_backend';
+
+import {SearchWiki} from './wikisearch.service';
+
+const mockResponse = {
+  "batchcomplete": "",
+  "continue": {
+    "sroffset": 10,
+    "continue": "-||"
+  },
+  "query": {
+    "searchinfo": {
+      "totalhits": 36853
+    },
+    "search": [{
+      "ns": 0,
+      "title": "Stuff",
+      "snippet": "<span></span>",
+      "size": 1906,
+      "wordcount": 204,
+      "timestamp": "2016-06-10T17:25:36Z"
+    }]
+  }
+};
+
+describe('Wikipedia search service', () => {
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      imports: [HttpModule],
+      providers: [
+        {
+          provide: XHRBackend,
+          useClass: MockBackend
+        },
+        SearchWiki
+      ]
+    });
+  });
+
+  it('should get search results', fakeAsync(
+    inject([
+      XHRBackend,
+      SearchWiki
+    ], (mockBackend, searchWiki) => {
+
+      const expectedUrl = 'https://en.wikipedia.org/w/api.php?' +
+        'action=query&list=search&srsearch=Angular';
+
+      mockBackend.connections.subscribe(
+        (connection: MockConnection) => {
+          expect(connection.request.url).toBe(expectedUrl);
+
+          connection.mockRespond(new Response(
+            new ResponseOptions({ body: mockResponse })
+          ));
+        });
+
+      searchWiki.search('Angular')
+        .subscribe(res => {
+          expect(res).toEqual(mockResponse);
+        });
+    })
+  ));
+
+});
+```
+[View Example](http://plnkr.co/edit/k5bQN2YuFMkNXrNC2Zo3?p=preview)
+
+We use `inject` to inject the `SearchWiki` service and the `MockBackend` into our test. We then wrap our entire test with a call to `fakeAsync`, which will be used to control the asynchronous behavior of the `SearchWiki` service for testing.
+
+Next, we `subscribe` to any incoming connections from our back-end. This gives us access to an object `MockConnection`, which allows us to configure the response we want to send out from our back-end, as well as test any incoming requests from the service we're testing.
+
+In our example, we want to verify that the `SearchWiki`'s search method makes a GET request to the correct URL. This is accomplished by looking at the request object we get when our `SearchWiki` service makes a connection to our mock back-end. Analyzing the `request.url` property we can see if its value is what we expect it to be. Here we are only checking the URL, but in other scenarios we can see if certain headers have been set, or if certain POST data has been sent.
```

Comment:
> we want to verify that the `SearchWiki`'s search method makes a GET request to the correct URL

In the test file we are verifying the request URL but we are not checking the method. Maybe we should add an additional expectation there to be consistent:

```js
expect(connection.request.method).toBe('GET');
expect(connection.request.url).toBe(expectedUrl);
```